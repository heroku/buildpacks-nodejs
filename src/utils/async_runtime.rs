use std::sync::LazyLock;
use tokio::runtime::Runtime;

pub(crate) static ASYNC_RUNTIME: LazyLock<Runtime> = LazyLock::new(|| {
    tokio::runtime::Builder::new_multi_thread()
        .enable_io()
        .enable_time()
        .build()
        .expect("Should be able to construct the Async Runtime")
});
