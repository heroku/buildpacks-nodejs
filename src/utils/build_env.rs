use crate::{BuildpackBuildContext, BuildpackError};
use libcnb::Env;
use libcnb::data::layer::LayerName;
use libcnb::layer::UncachedLayerDefinition;
use libcnb::layer_env::{LayerEnv, ModificationBehavior, Scope};
use std::ffi::OsString;

pub(crate) fn set_default_env_var(
    context: &BuildpackBuildContext,
    env: &mut Env,
    name: impl Into<OsString>,
    value: impl Into<OsString>,
) -> Result<(), BuildpackError> {
    let name = name.into();
    let value = value.into();
    let layer_name = format!("build_default_{}", name.to_string_lossy().to_lowercase())
        .parse::<LayerName>()
        .expect("Layer name should be valid");

    let layer = context.uncached_layer(
        layer_name,
        UncachedLayerDefinition {
            build: true,
            launch: false,
        },
    )?;

    let mut layer_env = LayerEnv::new();
    layer_env.insert(Scope::Build, ModificationBehavior::Default, name, value);

    layer.write_env(layer_env)?;

    env.clone_from(&layer.read_env()?.apply(Scope::Build, env));

    Ok(())
}
