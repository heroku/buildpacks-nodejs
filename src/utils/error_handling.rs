use crate::BuildpackError;
use bullet_stream::{Print, style};
use indoc::formatdoc;
use std::fmt::{Display, Formatter};
use std::path::Path;

pub(crate) const BUILDPACK_NAME: &str = "Heroku Node.js buildpack";

const ISSUES_URL: &str = "https://github.com/heroku/buildpacks-nodejs/issues";

pub(crate) fn on_framework_error<E>(error: &E) -> ErrorMessage
where
    E: Display,
{
    let issues_url = style::url(ISSUES_URL);
    error_message()
        .error_type(ErrorType::Framework)
        .header(format!("{BUILDPACK_NAME} internal error"))
        .body(formatdoc! {"
            The framework used by this buildpack encountered an unexpected error.

            If you canâ€™t deploy to Heroku due to this issue, check the official Heroku Status page at \
            status.heroku.com for any ongoing incidents. After all incidents resolve, retry your build.

            Use the debug information above to troubleshoot and retry your build. If you think you found a \
            bug in the buildpack, reproduce the issue locally with a minimal example and file an issue here:
            {issues_url}
        "})
        .debug_info(error.to_string())
        .create()
}

#[bon::builder(finish_fn = create, on(String, into), state_mod(vis = "pub"))]
#[allow(clippy::needless_pass_by_value)]
pub(crate) fn error_message(
    header: String,
    body: String,
    error_type: ErrorType,
    debug_info: Option<String>,
) -> ErrorMessage {
    let mut message_parts = vec![header.trim().to_string(), body.trim().to_string()];
    let issues_url = style::url(ISSUES_URL);
    let pack = style::value("pack");
    let pack_url =
        style::url("https://buildpacks.io/docs/for-platform-operators/how-to/integrate-ci/pack/");

    match error_type {
        ErrorType::Framework => {}
        ErrorType::Internal => {
            message_parts.push(formatdoc! { "
                The causes for this error are unknown. We do not have suggestions for diagnosis or a \
                workaround at this time. You can help our understanding by sharing your buildpack log \
                and a description of the issue at:
                {issues_url}

                If you're able to reproduce the problem with an example application and the {pack} \
                build tool ({pack_url}), adding that information to the discussion will also help. Once \
                we have more information around the causes of this error we may update this message.
            "});
        }
        ErrorType::UserFacing(suggest_retry_build, suggest_submit_issue) => {
            if let SuggestRetryBuild::Yes = suggest_retry_build {
                message_parts.push(
                    formatdoc! { "
                        Use the debug information above to troubleshoot and retry your build.
                    "}
                    .trim()
                    .to_string(),
                );
            }

            if let SuggestSubmitIssue::Yes = suggest_submit_issue {
                message_parts.push(formatdoc! { "
                    If the issue persists and you think you found a bug in the buildpack, reproduce \
                    the issue locally with a minimal example. Open an issue in the buildpack's GitHub \
                    repository and include the details here:
                    {issues_url}
                "}.trim().to_string());
            }
        }
    }

    let message = message_parts.join("\n\n");

    ErrorMessage {
        debug_info,
        message,
    }
}

pub(crate) fn file_value(value: impl AsRef<Path>) -> String {
    style::value(value.as_ref().to_string_lossy())
}

#[derive(Debug)]
pub(crate) struct ErrorMessage {
    debug_info: Option<String>,
    message: String,
}

impl Display for ErrorMessage {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        let mut log = Print::new(vec![]).without_header();
        if let Some(debug_info) = &self.debug_info {
            log = log
                .bullet(style::important("Debug Info:"))
                .sub_bullet(debug_info)
                .done();
        }
        let output = log.error(&self.message);
        write!(f, "{}", String::from_utf8_lossy(&output))
    }
}

impl From<ErrorMessage> for BuildpackError {
    fn from(value: ErrorMessage) -> Self {
        libcnb::Error::BuildpackError(value)
    }
}

#[derive(Debug, PartialEq)]
pub(crate) enum ErrorType {
    Framework,
    Internal,
    UserFacing(SuggestRetryBuild, SuggestSubmitIssue),
}

#[derive(Debug, PartialEq)]
pub(crate) enum SuggestRetryBuild {
    Yes,
    No,
}

#[derive(Debug, PartialEq)]
pub(crate) enum SuggestSubmitIssue {
    Yes,
    No,
}

#[cfg(test)]
pub(crate) mod test_util {
    use crate::utils::error_handling::ErrorMessage;
    use bullet_stream::strip_ansi;
    use fun_run::{CmdError, CommandWithName};
    use insta::{assert_snapshot, with_settings};
    use std::path::PathBuf;
    use std::process::Command;
    use test_support::test_name;

    pub(crate) fn assert_error_snapshot(error: &ErrorMessage) {
        let error_message = strip_ansi(error.to_string());
        let test_name = test_name()
            .replace("::", "_")
            .replace("_tests", "")
            .replace("_test", "");
        let snapshot_path = std::env::var("CARGO_MANIFEST_DIR")
            .map(PathBuf::from)
            .expect(
                "The CARGO_MANIFEST_DIR should be automatically set by Cargo when running tests",
            )
            .join("src/__snapshots");
        with_settings!({
            prepend_module_to_snapshot => false,
            omit_expression => true,
            snapshot_path => snapshot_path,
        }, {
            assert_snapshot!(test_name, error_message);
        });
    }

    pub(crate) fn create_reqwest_error() -> reqwest_middleware::Error {
        tokio::runtime::Runtime::new().unwrap().block_on(async {
            reqwest_middleware::Error::Reqwest(
                reqwest::get("https://test/error").await.unwrap_err(),
            )
        })
    }

    pub(crate) fn create_json_error() -> serde_json::error::Error {
        serde_json::from_str::<serde_json::Value>(r#"{\n  "name":\n}"#).unwrap_err()
    }

    pub(crate) fn create_cmd_error(command: impl Into<String>) -> CmdError {
        Command::new("false")
            .named(command.into())
            .named_output()
            .unwrap_err()
    }
}
