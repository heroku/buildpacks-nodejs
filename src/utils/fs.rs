use std::os::unix::fs::PermissionsExt;
use std::path::Path;

pub(crate) fn symlink_executable(
    original: impl AsRef<Path>,
    link: impl AsRef<Path>,
) -> Result<(), std::io::Error> {
    // Get current permissions
    let metadata = std::fs::metadata(&original)?;
    let mut permissions = metadata.permissions();

    // Set the user, group, and other executable bits (0o111)
    let mode = permissions.mode() | 0o111;
    permissions.set_mode(mode);

    // Ensure the Yarn binary is executable
    std::fs::set_permissions(&original, permissions)?;

    std::os::unix::fs::symlink(original, link)
}
