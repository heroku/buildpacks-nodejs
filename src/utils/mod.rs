pub(crate) mod async_runtime;
pub(crate) mod build_env;
pub(crate) mod buildplan;
pub(crate) mod config;
pub(crate) mod download;
pub(crate) mod error_handling;
pub(crate) mod http;
pub(crate) mod npmjs_org;
pub(crate) mod package_json;
pub(crate) mod runtime_env;
pub(crate) mod vrs;
