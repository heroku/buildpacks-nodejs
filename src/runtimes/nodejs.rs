use crate::utils::vrs::{Requirement, Version};
use libherokubuildpack::inventory::Inventory;
use libherokubuildpack::inventory::artifact::Artifact;
use sha2::Sha256;
use std::sync::LazyLock;

pub(crate) static NODEJS_INVENTORY: LazyLock<NodejsInventory> = LazyLock::new(|| {
    toml::from_str(include_str!("../../inventory/nodejs.toml"))
        .expect("Inventory file should be valid")
});

// TODO: Requirement should capture the original requirement string for display purposes but it
//       doesn't (yet), so this wrapper type will have to do for now.
pub(crate) static DEFAULT_NODEJS_REQUIREMENT: LazyLock<DefaultNodeRequirement> =
    LazyLock::new(|| {
        let current_lts = "22.x";
        DefaultNodeRequirement {
            value: current_lts.to_string(),
            requirement: Requirement::parse(current_lts)
                .expect("Default Node.js version should be valid"),
        }
    });

pub(crate) struct DefaultNodeRequirement {
    pub(crate) value: String,
    pub(crate) requirement: Requirement,
}

pub(crate) type NodejsArtifact = Artifact<Version, Sha256, Option<()>>;

pub(crate) type NodejsInventory = Inventory<Version, Sha256, Option<()>>;
