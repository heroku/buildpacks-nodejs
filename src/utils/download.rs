use crate::utils;
use crate::utils::async_runtime::ASYNC_RUNTIME;
use async_compression::tokio::bufread::GzipDecoder;
use bullet_stream::global::print;
use futures::StreamExt;
use futures::stream::TryStreamExt;
use sha2::{Digest, Sha256};
use std::io;
use std::path::{Path, PathBuf};
use std::time::Duration;
use tokio::fs::File as AsyncFile;
use tokio::io::{AsyncWriteExt, BufReader as AsyncBufReader, copy as async_copy, sink};
use tokio_util::io::{InspectReader, StreamReader};

pub(crate) fn download_sync<'a, T>(downloader: T) -> Result<(), DownloaderError>
where
    T: Downloader<'a>,
{
    ASYNC_RUNTIME.block_on(async { download(downloader).await })
}

#[allow(clippy::too_many_lines)]
pub(crate) async fn download<'a, T>(downloader: T) -> Result<(), DownloaderError>
where
    T: Downloader<'a>,
{
    let response = utils::http::get(downloader.source_url())
        .connect_timeout(downloader.connect_timeout())
        .read_timeout(downloader.read_timeout())
        .max_retries(downloader.max_retries())
        .call()
        .await
        .map_err(|e| DownloaderError::Request {
            url: downloader.source_url().to_string(),
            source: e,
        })?;

    let timer = print::sub_start_timer("Downloading");

    let checksum_type = downloader.checksum_validator();

    #[allow(clippy::manual_map)]
    let mut hasher = match checksum_type {
        Some(ChecksumValidator::Sha256(_)) => Some(Sha256::new()),
        None => None,
    };

    // ensure the writer is dropped in this scope so the InspectReader is dropped and is no longer borrowing the hash digest
    {
        let mut inspectable_reader = AsyncBufReader::new(
            // the inspect reader lets us pipe this decompressed output to both the output file and the hash digest
            InspectReader::new(
                StreamReader::new(response.bytes_stream().map_err(io::Error::other)),
                |bytes| {
                    if let Some(hasher) = &mut hasher {
                        hasher.update(bytes);
                    }
                },
            ),
        );

        let create_write_error = |e| DownloaderError::Write {
            url: downloader.source_url().to_string(),
            destination: downloader.destination().to_path_buf(),
            source: e,
        };

        match downloader.extractor() {
            Some(Extractor::Gzip(options)) => {
                let mut reader = GzipDecoder::new(inspectable_reader);
                // Enable/Disable support for multistream gz files. In this mode, the reader expects the input to
                // be a sequence of individually gzipped data streams, each with its own header and trailer,
                // ending at EOF. This is standard behavior for gzip readers.
                reader.multiple_members(options.multiple_members);
                let mut tar_archive = tokio_tar::Archive::new(reader);
                let mut entries = tar_archive.entries().map_err(create_write_error)?;
                while let Some(entry) = entries.next().await {
                    let mut entry = entry.map_err(create_write_error)?;
                    let path = entry.path().map_err(create_write_error)?;

                    // Get the path components
                    let path_components: Vec<_> = path.components().collect();

                    // Skip if we don't have enough components after stripping
                    if path_components.len() <= options.strip_components {
                        async_copy(&mut entry, &mut sink())
                            .await
                            .map_err(create_write_error)?;
                        continue;
                    }

                    // Skip if the path contains '..' or is absolute
                    if path_components.iter().any(|c| {
                        matches!(
                            c,
                            std::path::Component::ParentDir | std::path::Component::RootDir
                        )
                    }) {
                        async_copy(&mut entry, &mut sink())
                            .await
                            .map_err(create_write_error)?;
                        continue;
                    }

                    // Build the stripped path
                    let stripped_path: PathBuf = path_components
                        .into_iter()
                        .skip(options.strip_components)
                        .collect();

                    // Skip empty paths
                    if stripped_path.components().count() == 0 {
                        async_copy(&mut entry, &mut sink())
                            .await
                            .map_err(create_write_error)?;
                        continue;
                    }

                    // Skip excluded paths
                    let exclude = &options.exclude;
                    if exclude(&stripped_path) {
                        async_copy(&mut entry, &mut sink())
                            .await
                            .map_err(create_write_error)?;
                        continue;
                    }

                    let entry_type = entry.header().entry_type();
                    let dest_path = downloader.destination().join(&stripped_path);

                    // Create parent directories for regular files and symlinks
                    if (entry_type.is_file()
                        || entry_type.is_symlink()
                        || entry_type.is_hard_link())
                        && let Some(parent) = dest_path.parent()
                    {
                        tokio::fs::create_dir_all(parent)
                            .await
                            .map_err(create_write_error)?;
                    }

                    entry.unpack(dest_path).await.map_err(create_write_error)?;
                }
            }
            None => {
                let mut writer = AsyncFile::create(downloader.destination())
                    .await
                    .map_err(create_write_error)?;
                async_copy(&mut inspectable_reader, &mut writer)
                    .await
                    .map_err(create_write_error)?;
                writer.flush().await.map_err(create_write_error)?;
            }
        }
    }

    timer.done();

    if let Some(hasher) = hasher.take() {
        match checksum_type {
            Some(ChecksumValidator::Sha256(expected_value)) => {
                let actual_checksum = hasher.finalize();
                if expected_value != actual_checksum.to_vec() {
                    Err(DownloaderError::ChecksumMismatch {
                        url: downloader.source_url().to_string(),
                        expected_checksum: hex::encode(expected_value),
                        actual_checksum: format!("{actual_checksum:x}"),
                    })?;
                }
            }
            None => {}
        }
    }

    Ok(())
}

pub(crate) trait Downloader<'a> {
    fn source_url(&self) -> &str;

    fn destination(&self) -> &Path;

    fn checksum_validator(&self) -> Option<ChecksumValidator<'a>> {
        None
    }

    fn extractor(&self) -> Option<Extractor> {
        None
    }

    fn connect_timeout(&self) -> Duration {
        crate::utils::http::DEFAULT_CONNECT_TIMEOUT
    }

    fn read_timeout(&self) -> Duration {
        crate::utils::http::DEFAULT_READ_TIMEOUT
    }

    fn max_retries(&self) -> u32 {
        crate::utils::http::DEFAULT_RETRIES
    }
}

#[derive(Debug)]
pub(crate) enum Extractor {
    Gzip(GzipOptions),
}

pub(crate) struct GzipOptions {
    pub(crate) multiple_members: bool,
    pub(crate) strip_components: usize,
    pub(crate) exclude: Box<dyn Fn(&Path) -> bool>,
}

impl std::fmt::Debug for GzipOptions {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("GzipOptions")
            .field("multiple_members", &self.multiple_members)
            .field("strip_components", &self.strip_components)
            .field("exclude", &"<closure>")
            .finish()
    }
}

impl Default for GzipOptions {
    fn default() -> Self {
        Self {
            multiple_members: true,
            strip_components: 0,
            exclude: Box::new(|_| false),
        }
    }
}

#[derive(Debug)]
pub(crate) enum ChecksumValidator<'a> {
    Sha256(&'a [u8]),
}

#[derive(Debug)]
pub(crate) enum DownloaderError {
    Request {
        url: String,
        source: utils::http::Error,
    },
    Write {
        url: String,
        destination: PathBuf,
        source: io::Error,
    },
    ChecksumMismatch {
        url: String,
        expected_checksum: String,
        actual_checksum: String,
    },
}
