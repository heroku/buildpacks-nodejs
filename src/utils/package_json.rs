use crate::utils::vrs::Requirement;
use serde::Deserialize;
use std::fs::File;
use std::io::BufReader;
use std::path::Path;
use thiserror::Error;

#[derive(Deserialize, Debug, Default, Clone)]
pub(crate) struct PackageJson {
    pub(crate) engines: Option<Engines>,
    pub(crate) scripts: Option<Scripts>,
}

#[derive(Deserialize, Debug, Default, Clone)]
pub(crate) struct Engines {
    pub(crate) yarn: Option<Requirement>,
}

#[derive(Deserialize, Debug, Default, Clone)]
pub(crate) struct Scripts {
    pub(crate) start: Option<String>,
    pub(crate) build: Option<String>,
    #[serde(rename = "heroku-prebuild")]
    pub(crate) heroku_prebuild: Option<String>,
    #[serde(rename = "heroku-build")]
    pub(crate) heroku_build: Option<String>,
    #[serde(rename = "heroku-postbuild")]
    pub(crate) heroku_postbuild: Option<String>,
}

#[derive(Error, Debug)]
pub(crate) enum PackageJsonError {
    #[error("Could not read package.json. {0}")]
    AccessError(std::io::Error),
    #[error("Could not parse package.json. {0}")]
    ParseError(serde_json::Error),
}

impl PackageJson {
    /// Reads a package.json from a path into a `PackageJson` struct.
    ///
    /// # Errors
    ///
    /// * Invalid/malformed JSON
    /// * Path does not exist or is unreadable
    /// * Version strings are invalid/malformed
    pub(crate) fn read<P: AsRef<Path>>(path: P) -> Result<Self, PackageJsonError> {
        let file = File::open(path).map_err(PackageJsonError::AccessError)?;
        let rdr = BufReader::new(file);
        serde_json::from_reader(rdr).map_err(PackageJsonError::ParseError)
    }

    #[must_use]
    /// Fetches the build scripts from a `PackageJson` and returns them in priority order
    pub(crate) fn build_scripts(&self) -> Vec<String> {
        let mut scripts = vec![];
        if let Some(s) = &self.scripts {
            if s.heroku_prebuild.is_some() {
                scripts.push("heroku-prebuild".to_owned());
            }
            if s.heroku_build.is_some() {
                scripts.push("heroku-build".to_owned());
            } else if s.build.is_some() {
                scripts.push("build".to_owned());
            }
            if s.heroku_postbuild.is_some() {
                scripts.push("heroku-postbuild".to_owned());
            }
        }
        scripts
    }

    #[must_use]
    /// Determines if a given `PackageJson` has a start script defined
    pub(crate) fn has_start_script(&self) -> bool {
        self.scripts
            .as_ref()
            .is_some_and(|scripts| scripts.start.is_some())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::io::Write;
    use tempfile::Builder;

    #[test]
    fn read_empty_package() {
        let mut f = Builder::new().tempfile().unwrap();
        write!(f, "{{ }}").unwrap();
        let pkg = PackageJson::read(f.path());
        assert!(pkg.is_ok());
    }

    #[test]
    fn read_valid_package() {
        let mut f = Builder::new().tempfile().unwrap();
        write!(f, "{{\"name\": \"foo\",\"version\": \"0.0.0\"}}").unwrap();
        let pkg = PackageJson::read(f.path());
        assert!(pkg.is_ok());
    }

    #[test]
    fn read_missing_package() {
        let res = PackageJson::read(Path::new("/over/there/package.json"));
        assert!(res.is_err());
        let err = res.unwrap_err().to_string();
        println!("{err}");
        assert!(err.contains("Could not read package.json"));
    }

    #[test]
    fn read_invalid_package() {
        let mut f = Builder::new().tempfile().unwrap();
        write!(f, "{{").unwrap();
        let res = PackageJson::read(f.path());
        assert!(res.is_err());
        let err = res.unwrap_err().to_string();
        println!("{err}");
        assert!(err.contains("Could not parse package.json"));
    }

    #[test]
    fn test_build_scripts_all() {
        let pkg_json = PackageJson {
            scripts: Some(Scripts {
                build: Some("echo 'build'".to_owned()),
                heroku_prebuild: Some("echo 'heroku-prebuild'".to_owned()),
                heroku_build: Some("echo 'build'".to_owned()),
                heroku_postbuild: Some("echo 'heroku-postbuild'".to_owned()),
                ..Scripts::default()
            }),
            ..PackageJson::default()
        };
        let build_scripts = pkg_json.build_scripts();

        assert_eq!("heroku-prebuild", build_scripts[0]);
        assert_eq!("heroku-build", build_scripts[1]);
        assert_eq!("heroku-postbuild", build_scripts[2]);
    }

    #[test]
    fn test_build_scripts_build_only() {
        let pkg_json = PackageJson {
            scripts: Some(Scripts {
                build: Some("echo 'build'".to_owned()),
                ..Scripts::default()
            }),
            ..PackageJson::default()
        };
        let build_scripts = pkg_json.build_scripts();

        assert_eq!("build", build_scripts[0]);
    }
}
