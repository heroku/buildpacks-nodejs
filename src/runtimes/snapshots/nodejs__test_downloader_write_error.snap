---
source: src/runtimes/nodejs.rs
---
- Debug Info:
  - test I/O error

! Failed to copy Node.js distribution contents
!
! An unexpected I/O error occurred while writing the contents of the Node.js distribution from https://nodejs.org/download/release/v23.6.0/node-v23.6.0-linux-arm64.tar.gz to the installation directory at `/layers/heroku_nodejs/nodejs`.
!
! Use the debug information above to troubleshoot and retry your build.
!
! If the issue persists and you think you found a bug in the buildpack, reproduce the issue locally with a minimal example. Open an issue in the buildpack's GitHub repository and include the details here:
! https://github.com/heroku/buildpacks-nodejs/issues
